tools/chpldoc/findUndocumentedSymbols.py

```diff
@@ -0,0 +1,387 @@
+        aggregate_name = get_simple_node_name(p)
+        name = f"{aggregate_name}.{name}"
+
+    if (mod := get_module(node.parent())) and mod.kind() != "implicit":
+        module_name = mod.name()
+        name = f"{module_name}.{name}"
+
+    return name
+
+
+def lookAhead(iterable, window=1):
+    items, nexts = itertools.tee(iterable, 2)
+    nexts = itertools.islice(nexts, window, None)
+    return itertools.zip_longest(items, nexts)
+
+
+"""
+main code
+"""
+
+
+class FindUndocumentedSymbols:
+    def __init__(
+        self,
+        ast: List[dyno.AstNode],
+        ignore_deprecated: bool = False,
+        ignore_unstable: bool = False,
+    ):
+        self.ast = ast
+        self.ignore_deprecated = ignore_deprecated
+        self.ignore_unstable = ignore_unstable
+
+    # (pattern, extra check fun)
+    documentable_symbol_patterns = {
+        "function": (
+            dyno.Function,
+            lambda node, match: isinstance(
+                node.parent(), (dyno.Module, dyno.AggregateDecl, dyno.Interface)
+            ),
+        ),
+        "module": (dyno.Module, lambda node, match: node.kind() != "implicit"),
+        "aggregate_decl": (
+            dyno.AggregateDecl,
+            lambda node, match: isinstance(
+                node.parent(), (dyno.Module, dyno.AggregateDecl)
+            ),
+        ),
+        "decl": (
+            dyno.Variable,
+            lambda node, match: isinstance(
+                node.parent(), (dyno.Module, dyno.AggregateDecl)
+            ),
+        ),
+        "multi_decl": (
+            dyno.MultiDecl,
+            lambda node, match: isinstance(
+                node.parent(), (dyno.Module, dyno.AggregateDecl)
+            ),
+        ),
+        "enum": (
+            dyno.Enum,
+            lambda node, match: isinstance(
+                node.parent(), (dyno.Module, dyno.AggregateDecl)
+            ),
+        ),
+        "enum_element": (dyno.EnumElement, None),
+        "interface": (
+            dyno.Interface,
+            lambda node, match: isinstance(
+                node.parent(), (dyno.Module, dyno.AggregateDecl)
+            ),
+        ),
+    }
+
+    def _get_previous_sibling(self, node: dyno.AstNode):
+        parent = node.parent()
+        if not parent:
+            for sib1, sib2 in lookAhead(self.ast):
+                if sib2 and node.unique_id() == sib2.unique_id():
+                    return sib1
+            return None
+
+        for sib1, sib2 in lookAhead(parent):
+            if sib2 and node.unique_id() == sib2.unique_id():
+                return sib1
+        return None
+
+    def get_documentable_symbols(self, root: dyno.AstNode) -> Generator:
+        for (
+            pat,
+            check_func,
+        ) in FindUndocumentedSymbols.documentable_symbol_patterns.values():
+            matches = [
+                m for m in each_matching(root, pat) if not check_func or check_func(*m)
+            ]
```

Comment:
Style suggestion -- not sure if Pythonic.
```suggestion
            matches = [ m for m in each_matching(root, pat)
                          if not check_func or check_func(*m) ]
```

Reply:
I ran the code through pythons `black` prior to committing to format this, I'd rather leave it